data/tutorials/language/1ms_01_functors.md

```diff
@@ -8,183 +8,381 @@ category: "Module System"
 
 # Functors
 
-Functors are probably one of the most complex features of OCaml, but you don't
-have to use them extensively to be a successful OCaml programmer. Actually,
-you may never have to define a functor yourself, but you will surely encounter
-them in the standard library. They are the only way of using the Set and Map
-modules, but using them is not so difficult.
+## Introduction
 
-## What Are Functors and Why Do We Need Them?
+Learning goals:
+- How to use a functor
+- How to write a functor
+- When to use a functor, some cases
 
-A functor is a module that is parametrised by another module, just like a
-function is a value which is parametrised by other values, the arguments.
+A functor is just a parametrized module.
 
-It allows one to parametrise a type by a value, which is not possible directly
-in OCaml without functors. For example, we can define a functor that takes an
-`int n` and returns a collection of array operations that work exclusively on
-arrays of length `n`. If by mistake the programmer passes a regular array to one
-of those functions, it will result in a compilation error. If we were not using
-this functor but the standard array type, the compiler would not be able to
-detect the error, and we would get a runtime error at some undetermined date in
-the future, which is much worse.
+**Prerequisites**: Transitive closure leading to modules.
 
-## Using an Existing Functor
+## Project Setup
 
-The standard library defines a `Set` module, which provides a `Make` functor.
-This functor takes one argument, which is a module that provides (at least) two
-things: the type of elements, given as `t` and the comparison function given as
-`compare`. The point of the functor is to ensure that the same comparison
-function will always be used, even if the programmer makes a mistake.
+This tutorial uses the [Dune](https://dune.build) build tool. Make sure to have installed version 3.7 or later. We start by creating a fresh project. We need a folder named `funkt` with files `dune-project`, `dune`, `funkt.opam` and `funkt.ml`, the latter two are created empty.
+```shell
+$ mkdir funkt; cd funkt
 
-For example, if we want to use sets of `ints`, we would do this:
+$ touch funkt.opam funkt.ml
+```
+
+**`dune-project`**
+```lisp
+(lang dune 3.7)
+```
+
+**`dune`**
+```lisp
+(executable
+  (name funkt)
+  (public_name funkt)
+  (libraries str))
+```
+
+Check this works using the `dune exec funkt` command, it shouldn't do anything (the empty file is valid OCaml syntax) but it shouldn't fail either. The stanza `libraries str` will be used later.
+
+## Using an Existing Functor: `Set.Make`
+
+The standard library contains a [`Set`](/api/Set.html) module providing a data structure that allows operations like union and intersection. To use the provided type and its associated [functions](/api/Set.S.html), it is required to use the functor provided by `Set`. For reference only, here is a shortened version of the interface of `Set`.
```

Comment:
I think I'm missing _why_ we need to use the functor provided by `Set` here – even a small justification for how we want to _reuse_ the functionality of a Set but with a custom `compare` operation.

Reply:
What do you think of something like this:
> Most of the implementations of the set operations must use a comparison function. Using the default `Stdlib.compare` would make it impossible to use a user-defined comparison algorithm. Passing the comparaison as higher-order parameter (like `Array.sort` does for instance) would add a lot of boilerplate. Providing set operations as a functor allows to specify the comparison function only once.

Reply:
Yes, and this would pave the way for the arguments for functors as a tool for programming-in-the-large! I dig it.

Things like "avoiding repetitive code", "helping establish patterns", or "customizing a lot of code quickly".

Reply:
Text added as a note below